pkg/requestreply/cesql.go

```diff
@@ -27,45 +30,59 @@ import (
 	"knative.dev/pkg/logging"
+	"knative.dev/pkg/system"
 )
 
 var once = sync.Once{}
 
 // RegisterCESQLVerifyCorrelationIdFilter registers a new function with the CESQL runtime to verify correlation ids
-// The function signature is KN_VERIFY_CORRELATION_ID(correlation_id, namespace, secret_names...)
+// The function signature is KN_VERIFY_CORRELATION_ID(correlation_id, name, namespace, secret_name, pod_idx, replica_count)
 func RegisterCESQLVerifyCorrelationIdFilter(ctx context.Context) error {
 	secretInformer := secretinformer.Get(ctx)
 	logger := logging.FromContext(ctx)
 
 	var verifyCorrelationId = cefn.NewFunction(
 		"KN_VERIFY_CORRELATION_ID",
-		[]cesql.Type{cesql.StringType, cesql.StringType},
-		cesql.TypePtr(cesql.StringType),
+		[]cesql.Type{cesql.StringType, cesql.StringType, cesql.StringType, cesql.StringType, cesql.IntegerType, cesql.IntegerType},
+		nil,
 		cesql.BooleanType,
 		func(e cloudevents.Event, args []interface{}) (interface{}, error) {
-			replyId, namespace := args[0].(string), args[1].(string)
-			secretNames := args[2:]
-
-			for _, secretName := range secretNames {
-				secret, err := secretInformer.Lister().Secrets(namespace).Get(secretName.(string))
-				if err != nil {
-					logger.Warnf("failed to get secret %s in namespace %s: %s", secretName.(string), namespace, err.Error())
-					continue
-				}
+			replyId, name, namespace, secretName, podIdx, replicaCount := args[0].(string), args[1].(string), args[2].(string), args[3].(string), args[4].(int32), args[5].(int32)
+			logger.Info("handling CESQL function for KN_VERIFY_CORRELATION_ID")
 
-				aesKey, ok := secret.Data["key"]
-				if !ok {
-					logger.Warnf("no key in secret %s in namespace %s", secretName.(string), namespace)
-					continue
-				}
+			parts := strings.Split(replyId, ":")
+			if len(parts) != 3 {
+				return false, nil
+			}
 
-				validId, err := VerifyReplyId(replyId, aesKey)
-				if err != nil {
-					logger.Warnf("failed to verify replyid for event: %s", err)
-				}
+			idx, err := strconv.ParseInt(parts[2], 10, 32)
+			if err != nil {
+				logger.Warnf("failed to parse correlation id index to int")
+				return false, nil
+			}
+			if int32(idx)%replicaCount != podIdx {
+				return false, nil
+			}
+
+			secret, err := secretInformer.Lister().Secrets(system.Namespace()).Get(secretName)
+			if err != nil {
+				logger.Errorf("failed to get secret %s in namespace %s: %s", secretName, namespace, err.Error())
+				return false, nil
+			}
+
+			for keyName, aesKey := range secret.Data {
+				logger.Infof("trying key %s for namespace=%s name=%s", keyName, namespace, name)
+				if strings.HasPrefix(keyName, fmt.Sprintf("%s.%s.", namespace, name)) {
+					validId, err := VerifyReplyId(replyId, aesKey)
+					if err != nil {
+						logger.Warnf("failed to verify replyid for event: %s", err)
+					}
+
+					logger.Infof("finished key validation: %b", validId)
```

Comment:
@Cali0707 this is a more general question, but could you elaborate a bit on the need for the secrets and the replyid verification?

Reply:
yeah, but then they would at least have a trigger from the broker...

I think the use case here is to avoid having a pod that is not subscribed to the broker sending events into the broker that could act as a reply

Reply:
Ok. Thanks